File: src/fmj/characters/FightingSprite.kt
package fmj.characters

import fmj.lib.DatLib
import fmj.lib.ResImage
import graphics.Canvas

class FightingSprite(resType: DatLib.ResType, index: Int) {

    private var mImage: ResImage? = null

    /**
     *
     * @param i > 0
     */
    var currentFrame = 1

    /** 在战斗场景中的屏幕坐标 */
    var combatX: Int = 0
        private set
    var combatY: Int = 0
        private set

    val width: Int
        get() = mImage!!.width

    val height: Int
        get() = mImage!!.height

    val frameCnt: Int
        get() = mImage!!.number

    init {
        if (resType == DatLib.ResType.ACP) { // 怪物的
            mImage = DatLib.Companion.getRes(DatLib.ResType.ACP, 3, index) as ResImage
        } else if (resType == DatLib.ResType.PIC) { // 玩家角色的
            mImage = DatLib.Companion.getRes(DatLib.ResType.PIC, 3, index) as ResImage
        } else {
            throw Error("resType Error. resType=$resType")
        }
    }

    fun draw(canvas: Canvas) {
        mImage!!.draw(canvas, currentFrame, combatX - mImage!!.width / 2,
                combatY - mImage!!.height / 2)
    }

    fun draw(canvas: Canvas, x: Int, y: Int) {
        mImage!!.draw(canvas, currentFrame, x, y)
    }

    fun setCombatPos(x: Int, y: Int) {
        combatX = x
        combatY = y
    }

    fun move(dx: Int, dy: Int) {
        combatX += dx
        combatY += dy
    }
}

File: packages/core/src/characters/fighting-sprite.ts
import { type ResImage } from '@/lib/res-image';

// 玩家 PIC 3 战斗图的固定帧约定；怪物 ACP 3 只保证 Idle，其余按资源帧数播放。
export const PlayerFightingFrame = {
  Idle: 1,
  GuardedHit: 8,
  Defend: 9,
  Hit: 10,
  Weak: 11,
  Dead: 12,
} as const;
export type PlayerFightingFrame = (typeof PlayerFightingFrame)[keyof typeof PlayerFightingFrame];

export const PlayerFightingFrameRange = {
  PhysicalAttack: { start: 1, end: 5 },
  MagicCast: { start: 6, end: 8 },
} as const;
export type PlayerFightingFrameRange = (typeof PlayerFightingFrameRange)[keyof typeof PlayerFightingFrameRange];

export const MonsterFightingFrame = {
  Idle: 1,
} as const;
export type MonsterFightingFrame = (typeof MonsterFightingFrame)[keyof typeof MonsterFightingFrame];

// 战斗图在角色资源解析时绑定图片
export class FightingSprite {
  currentFrame: number = PlayerFightingFrame.Idle;
  combatX = 0;
  combatY = 0;

  constructor(readonly image: ResImage) {}

  get width(): number {
    return this.image.width;
  }

  get height(): number {
    return this.image.height;
  }

  get frameCount(): number {
    return this.image.number;
  }

  setCombatPos(x: number, y: number): void {
    this.combatX = x;
    this.combatY = y;
  }

  move(dx: number, dy: number): void {
    this.combatX += dx;
    this.combatY += dy;
  }
}
